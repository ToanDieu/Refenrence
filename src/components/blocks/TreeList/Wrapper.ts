import { css } from "@emotion/core";
import styled from "@emotion/styled";

export interface WrapperProps {
  edgeLength?: number;
  edgeThick?: number;
  edgeColor?: string;
}

const mixin = ({
  edgeLength = 15,
  edgeThick = 1,
  edgeColor = "#000"
}: WrapperProps) => css`
  & ul {
    margin: 0;
    padding: 0;

    list-style: none;
    display: flex;
    justify-content: center;
    position: relative;

    li {
      position: relative;
    }

    ul {
      padding-top: ${edgeLength}px;

      &::before {
        content: "";
        position: absolute;
        top: 0;
        left: calc(50% - ${edgeThick / 2}px);
        border-left: ${edgeThick}px solid ${edgeColor};
        height: ${edgeLength}px;
      }

      li {
        padding-top: ${edgeLength}px;

        &::before,
        &::after {
          content: "";
          position: absolute;
          top: 0;
          right: calc(50% - ${edgeThick / 2}px);
          border-top: ${edgeThick}px solid ${edgeColor};
          width: calc(50% + ${edgeThick / 2}px);
          height: ${edgeLength}px;
        }

        &::after {
          right: auto;
          left: calc(50% - ${edgeThick / 2}px);
          border-left: ${edgeThick}px solid ${edgeColor};
        }

        &:first-of-type::before,
        &:last-of-type::after {
          display: none;
        }

        &:last-of-type::before {
          border-right: ${edgeThick}px solid ${edgeColor};
        }

        &:only-of-type::before {
          display: block;
          border: none;
          left: calc(50% - ${edgeThick / 2}px);
          border-left: ${edgeThick}px solid ${edgeColor};
        }
      }
    }
  }
`;

export default styled.div<WrapperProps>`
  ${mixin}
`;
