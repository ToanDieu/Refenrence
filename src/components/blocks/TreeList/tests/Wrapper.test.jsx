import React from "react";
import { render } from "react-testing-library";

import Wrapper from "../Wrapper";

describe("<Wrapper />", () => {
  it("should match styles in snapshot", () => {
    const { container } = render(<Wrapper />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should match styles in snapshot with different props", () => {
    const { container } = render(<Wrapper edgeLength={8} edgeThick={3} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render an <div> tag", () => {
    const { container } = render(<Wrapper />);
    expect(container.firstElementChild.tagName).toEqual("DIV");
  });

  it("should have a class attribute", () => {
    const { container } = render(<Wrapper />);
    const element = container.firstElementChild;
    expect(element.hasAttribute("class")).toBe(true);
  });

  it("should adopt a valid attribute", () => {
    const id = "test";
    const { container } = render(<Wrapper id={id} />);
    const element = container.firstElementChild;
    expect(element.id).toEqual(id);
  });

  it("should not adopt an invalid attribute", () => {
    const { container } = render(<Wrapper attribute="test" />);
    const element = container.firstElementChild;
    expect(element.hasAttribute("attribute")).toBe(false);
  });
});
