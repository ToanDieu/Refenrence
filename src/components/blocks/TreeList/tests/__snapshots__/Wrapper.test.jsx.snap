// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Wrapper /> should match styles in snapshot 1`] = `
.emotion-0 ul {
  margin: 0;
  padding: 0;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.emotion-0 ul li {
  position: relative;
}

.emotion-0 ul ul {
  padding-top: 15px;
}

.emotion-0 ul ul::before {
  content: "";
  position: absolute;
  top: 0;
  left: calc(50% - 0.5px);
  border-left: 1px solid #000;
  height: 15px;
}

.emotion-0 ul ul li {
  padding-top: 15px;
}

.emotion-0 ul ul li::before,
.emotion-0 ul ul li::after {
  content: "";
  position: absolute;
  top: 0;
  right: calc(50% - 0.5px);
  border-top: 1px solid #000;
  width: calc(50% + 0.5px);
  height: 15px;
}

.emotion-0 ul ul li::after {
  right: auto;
  left: calc(50% - 0.5px);
  border-left: 1px solid #000;
}

.emotion-0 ul ul li:first-of-type::before,
.emotion-0 ul ul li:last-of-type::after {
  display: none;
}

.emotion-0 ul ul li:last-of-type::before {
  border-right: 1px solid #000;
}

.emotion-0 ul ul li:only-of-type::before {
  display: block;
  border: none;
  left: calc(50% - 0.5px);
  border-left: 1px solid #000;
}

<div
  class="emotion-0"
/>
`;

exports[`<Wrapper /> should match styles in snapshot with different props 1`] = `
.emotion-0 ul {
  margin: 0;
  padding: 0;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.emotion-0 ul li {
  position: relative;
}

.emotion-0 ul ul {
  padding-top: 8px;
}

.emotion-0 ul ul::before {
  content: "";
  position: absolute;
  top: 0;
  left: calc(50% - 1.5px);
  border-left: 3px solid #000;
  height: 8px;
}

.emotion-0 ul ul li {
  padding-top: 8px;
}

.emotion-0 ul ul li::before,
.emotion-0 ul ul li::after {
  content: "";
  position: absolute;
  top: 0;
  right: calc(50% - 1.5px);
  border-top: 3px solid #000;
  width: calc(50% + 1.5px);
  height: 8px;
}

.emotion-0 ul ul li::after {
  right: auto;
  left: calc(50% - 1.5px);
  border-left: 3px solid #000;
}

.emotion-0 ul ul li:first-of-type::before,
.emotion-0 ul ul li:last-of-type::after {
  display: none;
}

.emotion-0 ul ul li:last-of-type::before {
  border-right: 3px solid #000;
}

.emotion-0 ul ul li:only-of-type::before {
  display: block;
  border: none;
  left: calc(50% - 1.5px);
  border-left: 3px solid #000;
}

<div
  class="emotion-0"
/>
`;
